py/dml/traits.py

```diff
@@ -759,6 +752,40 @@ def __lt__(self, other):
     def type(self):
         return TTrait(self)
 
+    def typecheck_methods(self):
+        for (_, inp, outp, _, _, _, _) in self.vtable_methods.values():
+            for (_, t) in inp + outp:
+                try:
+                    check_named_types(t)
+                except DMLError as e:
+                    report(e)
+
+        for sm in self.method_impls.values():
+            # To avoid duplicating error messages
+            bad = False
+            if sm.name not in self.vtable_methods:
+                for (_, t) in sm.inp + sm.outp:
+                    try:
+                        check_named_types(t)
+                    except DMLError as e:
+                        bad = True
+                        report(e)
+
+            if not bad and sm.name in self.ancestor_vtables:
+                try:
+                    typecheck_method_override(
+                        (sm.site, sm.inp, sm.outp, sm.throws, sm.independent,
+                         sm.startup, sm.memoized),
+                        self.ancestor_vtables[sm.name].vtable_methods[sm.name])
+                except DMLError as e:
+                    report(e)
+
+        # We don't attempt to purge the bad methods. Subtemplates could
+        # be referencing them, and there is no good way to purge those
+        # references. This doesn't seem to be too big of a deal; DMLC stops
```

Comment:
when we have reported an error, the ambition is to cover up to the point that we dont ICE. it is fine e.g. if the fallback causes the c backend to generate broken / inconsistent code, or if it sometimes reports weird secondary errors.

Reply:
Indeed. This comment is because, before this PR, the bad shared methods _used_ to be pruned (and still are if they fail on something other than the signature.) But now that the typechecking happens so late that's no longer feasible (or, at least, not easy.) From my (admittedly shallow) investigation it doesn't _seem_ to be provoking ICE:s; however, the fact that the overriding method will linger about with an incorrect signature means methods can invoke it, which can lead to hilarity. That's the downside.